---
title: "Are differentially methylated regions within genes associated with mantle cell lymphoma?"
author: "Pascal Lafrenz, Mari Hambardzumyan, Lea Herzel, Franziska Lam"
output:
  pdf_document: default
  word_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

---

The alterations in cancer cell DNA methylation pattern have long been detected (Vogelstein and Feinberg, 1983). In many cancer types methylation level has also been shown to have effects on gene deregulation,  (Melnick, 2013). But not much research has been done associated with methylation in *mantle cell lymphoma* (MCL) cells. Therefore in the following project we want to address the following question.

# Research question

Are differentially methylated regions associated with mantle cell lymphoma?

For that purpose we received a data set and an annotation file. Below is a short summary of important key points of the data set.

# Data summary

* Bisulfite-Seq data for 10 patients/samples (5 male, 5 female)
* 5 healthy patients, 5 with MCL
* Extracted DNA is from B-cells
* 56175 genes across all samples
* coverage and beta-values for each gene
* Length of each gene

Researching on MCL we found a range of genes which have been linked with this specific cancer, some of which include SOX11, p53, CCND1 etc (Queir√≥s et al., 2016).

To address the research question we broke down the project into 5 parts.

# Project steps



|1. Data processing  | 2. Data normalization and visualization | 3. Data reduction  | 4. Regression and interpretation    | 5. Annotate and interpret results
------------- | ------------- | ------------- | ------------- | -------------
Remove x and y chromosome genes  | Transform beta-values into homoscedastic M-values  | Dimensionality reduction, e.g PCA  | Apply t-test        |  
Split data in 4 data frames (cancer vs. control, beta-values and coverage)  | Add a column for mean, median and SD for better comparisons  | Apply clustering method and visualize, e.g k-means         | Apply logistic regression        | 
Set a threshold and examine variance for reliable coverage, replace unreliable beta-values with NA | Visualization through plots | 
Separately inspect genes of interest for MCL | Identify batch effect magnitude in the data set
Set a threshold and remove genes under/above threshold, imputate remaining NA's | 



## Part 1: Data processing

For now, we want to single out the gene data frame of the patients to inspect it separately. We will examine only the gene data frame and neglect the data from tiling, promoters and cpg-islands because they are irrelevant to answer our research question.

```{r}
Samples <- readRDS(file = "Mantle-Bcell_list.RDS.gz")
Gene_data_frame <- Samples$genes
dim(Gene_data_frame)
```

There are coverage and beta-values of 56175 genes across all 10 patients. After the entire data clean-up we still want to retain around 90% of that information (~500000 genes).
$$~$$
We are starting the data clean-up with the sex chromosomes. In our case we have different numbers of female/male patients in healthy/cancer sample groups. As in female X chromosome inactivation  the majority of the genes on one X chromosome are silenced, this would lead to significant differences in the methylation profiles between female/male patients (Sharp et al., 2011). Therefore, we remove gene data of sex chromosomes. 
```{r}
Gene_data_frame_x_y <- 
  Gene_data_frame[-which(Gene_data_frame$Chromosome == "chrX"),]
Gene_data_frame_x_y <- 
  Gene_data_frame_x_y[-which(Gene_data_frame_x_y$Chromosome == "chrY"),]
```

Now we want to go through the coverage values separately to remove any unreliable values, outliers or possible systematic errors. First we separated the coverage values of healthy and cancer patients into 2 data frames. We did the same separation into 2 data frames with the beta-values because we will need these later.

```{r}
healthy_coverage <- Gene_data_frame_x_y[, 21:25]
cancer_coverage <- Gene_data_frame_x_y[, 26:30]
healthy_beta_values <- Gene_data_frame_x_y[, 11:15]
cancer_beta_values <- Gene_data_frame_x_y[, 16:20]
```

Then we created a histogram with mean coverage values for each gene on a log scale and added 10 quantiles from 10 % to 100 %.

```{r}
mean_cancer_coverage <- rowMeans(cancer_coverage)
hist(
  log10(mean_cancer_coverage),
  breaks = "fd",
  main = "Cancer coverage: Mean frequency",
  xlab = "Common logarithm of coverages",
  col = c("black"),
  border = "white"
)
abline(v = log10(quantile(
  mean_cancer_coverage,
  probs = seq(0, 1, 0.1),
  na.rm = TRUE
)),
col = colors(256),
lwd = 2)


mean_healthy_coverage <- rowMeans(healthy_coverage)
hist(
  log10(mean_healthy_coverage),
  breaks = "fd",
  main = "Healthy coverage: Mean frequency",
  xlab = "Common logarithm of coverages",
  col = c("black"),
  border = "white"
)
abline(v = log10(quantile(
  mean_healthy_coverage,
  probs = seq(0, 1, 0.1),
  na.rm = TRUE
)),
col = colors(256),
lwd = 2)
```

Then to check how widely spread are the coverages across the patients, we calculated the standard deviations (SD) of each gene across the patients.

```{r}
sd_cancer_coverage <- apply(cancer_coverage, 1, sd)
sd_healthy_coverage <- apply(healthy_coverage, 1, sd)
```

Consequently, we plotted a histogram with the SD values for each data frame.

```{r}
hist(log10(sd_cancer_coverage),
     breaks = "fd",
     main = "Cancer coverage: SD frequency",
     xlab = "Common logarithm of coverages",
     col = "indianred2"
)
hist(log10(sd_healthy_coverage),
     breaks = "fd",
     col = "seagreen2",
     main = "Healthy coverage: SD frequency",
     xlab = "Common logarithm of coverages",
)
```

By setting the lower coverage threshold to 5 % and the upper to 99.9 %, we fulfilled our initial requirement of retaining around 90 % of the information. 

```{r}
#cancer coverages: lower boundary
threshold <-
  quantile(mean_cancer_coverage,
           probs = seq(0.05, 0.05, 0.05),
           na.rm = TRUE)

#cancer coverages: upper boundary
threshold2 <-
  quantile(mean_cancer_coverage,
           probs = seq(0.95, 0.95, 0.05),
           na.rm = TRUE)

#healthy coverages: lower boundary
threshold3 <-
  quantile(mean_healthy_coverage,
           probs = seq(0.05, 0.05, 0.05),
           na.rm = TRUE)

#healthy coverages: upper boundary
threshold4 <-
  quantile(mean_healthy_coverage,
           probs = seq(0.95, 0.95, 0.05),
           na.rm = TRUE)
```

Beta-values with very low or extreme high coverages are not reliable. Therefore with the help of the predefined coverage threshold for each patient we will set the beta values of the genes which didn't fulfill the coverage requirement to NA.

```{r}
for (i in 1:nrow(cancer_coverage)) {
  for (j in 1:ncol(cancer_coverage)) {
    if (cancer_coverage[i, j] <= threshold) {
      cancer_coverage[i, j] <- 0
    }
    
    if (cancer_coverage[i, j] >= threshold2) {
      cancer_coverage[i, j] <- 0
    }
    
    if (cancer_coverage[i, j] == 0) {
      cancer_coverage[i, j] <- NA
      cancer_beta_values[i, j] <- NA
      
    }
  }
}

for (i in 1:nrow(healthy_coverage)) {
  for (j in 1:ncol(healthy_coverage)) {
    if (healthy_coverage[i, j] <= threshold3) {
      healthy_coverage[i, j] <- 0
    }
    
    if (healthy_coverage[i, j] >= threshold4) {
      healthy_coverage[i, j] <- 0
    }
    
    if (healthy_coverage[i, j] == 0) {
      healthy_coverage[i, j] <- NA
      healthy_beta_values[i, j] <- NA
      
    }
  }
}
```

?Erfolg der Datenbereinigung mit Histogrammen belegen/veranschaulichen? (211-228)

```{r}

```

Now we needed to deal with NA's in the beta-values. We decided that we want to remove each gene with 3 or more NA's in 'healthy_beta_values' or 'cancer_beta_values', because if there are not enough values our results will be not meaningful. To do so we added a new column with the number of NA's in each gene. 

```{r}
healthy_beta_values$Number_of_NA <-
  rowSums(is.na(healthy_beta_values))
cancer_beta_values$Number_of_NA <-
  rowSums(is.na(cancer_beta_values))
```

?Visualize number of NA's in a histogram? 

```{r}

```

Then we removed the genes with to much missing values. Here it was important to remove the genes in both data frames, because a gene in only one data frame without values to compare is meaningless. So we checked if the rownames are the same in both data frames after the deletion. 
```{r}
healthy_beta_values <-
  healthy_beta_values[!(healthy_beta_values$`Number_of_NA` > 2),]
cancer_beta_values <-
  cancer_beta_values[!(cancer_beta_values$`Number_of_NA` > 2),]

cancer_beta_values <-
  healthy_beta_values[!(healthy_beta_values$`Number_of_NA` > 2),]
healthy_beta_values <-
  cancer_beta_values[!(cancer_beta_values$`Number_of_NA` > 2),]
sum(rownames(healthy_beta_values) == rownames(cancer_beta_values))
```

We removed the columns with the NA's-values because we did not need them anymore. 

```{r}
healthy_beta_values <-
  healthy_beta_values[,-which(colnames(healthy_beta_values)  %in%  c('Number_of_NA'))]
cancer_beta_values <-
  cancer_beta_values[,-which(colnames(cancer_beta_values) %in% c('Number_of_NA'))]
```

We can not keep the remaining NA's in the beta-values because this will cause errors in our downstream analysis. But we also did not want to delete the genes, so we decided to replace the NA's with the mean value of the respective gene.

?genauer auf transposed data frames eingehen?

```{r}
transposed_healthy_beta_values <- t(healthy_beta_values)
transposed_cancer_beta_values <- t(cancer_beta_values)
for (i in 1:ncol(transposed_healthy_beta_values)) {
  transposed_healthy_beta_values[is.na(transposed_healthy_beta_values[, i]), i] <-
    mean(transposed_healthy_beta_values[, i], na.rm = TRUE)
}

for (i in 1:ncol(transposed_cancer_beta_values)) {
  transposed_cancer_beta_values[is.na(transposed_cancer_beta_values[, i]), i] <-
    mean(transposed_cancer_beta_values[, i], na.rm = TRUE)
}
healthy_beta_values <- data.frame(t(transposed_healthy_beta_values))
cancer_beta_values <- data.frame(t(transposed_cancer_beta_values))
```

We proofed that there are no more NA's left in the 2 data frames.

```{r}
sum(is.na(transposed_healthy_beta_values))
sum(is.na(transposed_cancer_beta_values))
```

There are some important genes for MCL which have ofen a different methylation pattern in comparison to heatlhy b-cells. These genes are:
*SOX11	ENSG00000176887	
*NR2F2	ENSG00000185551	
*p53  	ENSG00000141510	
*AHR  	ENSG00000106546		
*ROBO1	ENSG00000169855		
*SOX9 	ENSG00000125398		
*HOXA9	ENSG00000078399		
*CDH1 	ENSG00000039068		
*CDC14B	ENSG00000081377		
*FOXC1	ENSG00000054598	
*G0S2	  ENSG00000123689		
*GPX3	  ENSG00000211445		
*LGALS3	ENSG00000131981		
*MAL  	ENSG00000172005		
*NPTX2	ENSG00000106236		 
*PAX6	  ENSG00000007372		
*TFPI2	ENSG00000105825		
*THEM4	ENSG00000159445		
*TWIST1	ENSG00000122691		

To be sure we did not removed these genes in our data processing we tested if they are still all in our data frames. Because each of the 2 data frames contains the same genes it is enough to check if they are part of one of them.

```{r}
important_genes <-
  data.frame(
    c(
      "ENSG00000176887",
      "ENSG00000185551",
      "ENSG00000141510",
      "ENSG00000110092",
      "ENSG00000106546",
      "ENSG00000169855",
      "ENSG00000125398",
      "ENSG00000078399",
      "ENSG00000039068",
      "ENSG00000081377",
      "ENSG00000054598",
      "ENSG00000123689",
      "ENSG00000211445",
      "ENSG00000131981",
      "ENSG00000172005",
      "ENSG00000106236",
      "ENSG00000007372",
      "ENSG00000105825",
      "ENSG00000159445",
      "ENSG00000122691"
    )
  )
important_genes_in_data_set <- data.frame()
for (i in 1:nrow(important_genes)) {
  important_genes_in_data_set[i, ] <-
    cancer_beta_values[which(row.names(cancer_beta_values) == important_genes[i, ]), ]
}
```































# References

Sharp, A.J., Stathaki, E., Migliavacca, E., Brahmachary, M., Montgomery, S.B., Dupre, Y., and Antonarakis, S.E. (2011). DNA methylation profiles of human active and inactive X chromosomes. Genome Res 21, 1592-1600.